shell/browser/lib/shutdown_blocker_win.cc

```diff
@@ -0,0 +1,130 @@
+  BlockShutdown();
+}
+
+void ShutdownBlockerWin::BlockShutdown() {
+  if (blocking_) {
+    return;
+  }
+  blocking_ = true;
+  if (dedicated_message_loop_) {
+    LOG(INFO) << "unblocking shutdown on renderer process";
+    owner_thread_task_runner_ = base::ThreadTaskRunnerHandle::Get();
+    owner_thread_task_runner_->PostTask(
+        FROM_HERE, base::BindOnce(&ShutdownBlockerWin::MessageLoop,
+                                  base::Unretained(this)));
+  } else {
+    LOG(INFO) << "blocking shutdown on browser process";
+    CreateHiddenWindow();
+  }
+}
+
+void ShutdownBlockerWin::MessageLoop() {
+  CreateHiddenWindow();
+  MSG msg;
+  while (GetMessage(&msg, NULL, 0, 0)) {
+    TranslateMessage(&msg);
+    DispatchMessage(&msg);
+  }
+  DestroyHiddenWindow();
+}
+
+void ShutdownBlockerWin::CreateHiddenWindow() {
+  ATOM atom = RegisterWindowClass();
+  window_ = CreateWindow(MAKEINTATOM(atom), 0, WS_POPUP, 0, 0, 0, 0, 0, 0,
+                         instance_, 0);
+  gfx::CheckWindowCreated(window_, ::GetLastError());
+  gfx::SetWindowUserData(window_, this);
+}
+
+void ShutdownBlockerWin::DestroyHiddenWindow() {
+  DestroyWindow(window_);
+  UnregisterClass(kShutdownBlockerWinWindowClass, instance_);
+}
+
+ATOM ShutdownBlockerWin::RegisterWindowClass() {
+  // Register our window class
+  WNDCLASSEX window_class;
+  base::win::InitializeWindowClass(
+      kShutdownBlockerWinWindowClass,
+      &base::win::WrappedWindowProc<ShutdownBlockerWin::WndProc>, 0, 0, 0, NULL,
+      NULL, NULL, NULL, NULL, &window_class);
+  ATOM atom = RegisterClassEx(&window_class);
+  instance_ = window_class.hInstance;
+  return atom;
+}
+
+bool ShutdownBlockerWin::OnQueryEndSession() {
+  if (dedicated_message_loop_) {
+    owner_thread_task_runner_->PostTask(
+        FROM_HERE,
+        base::BindOnce(&ShutdownBlockerWin::OwnerThreadShutdownHandler,
+                       base::Unretained(this)));
+    return false;
+  }
+  return !should_shutdown_ || should_shutdown_.Run();
+}
+
+void ShutdownBlockerWin::OwnerThreadShutdownHandler() {
+  if (should_shutdown_) {
+    should_shutdown_.Run();
+  }
+}
+
+LRESULT CALLBACK ShutdownBlockerWin::WndProc(HWND hwnd,
+                                             UINT message,
+                                             WPARAM wparam,
+                                             LPARAM lparam) {
+  switch (message) {
+    case WM_QUERYENDSESSION: {
+      ShutdownBlockerWin* blocker = reinterpret_cast<ShutdownBlockerWin*>(
+          GetWindowLongPtr(hwnd, GWLP_USERDATA));
+      LOG(INFO) << "Received WM_QUERYENDSESSION"
+                << (!blocker ? ""
+                             : (blocker->dedicated_message_loop_
+                                    ? " on renderer process"
+                                    : " on browser process"));
+      if (blocker && !blocker->OnQueryEndSession()) {
+        LOG(INFO) << "Shutdown blocked";
+        ShutdownBlockReasonCreate(hwnd, L"Ensure a clean shutdown");
```

Comment:
Calling `ShutdownBlockReasonCreate()` inside the `WM_QUERYENDSESSION` handler allows us to notify our code to clean up and shutdown quickly, but we can also use it preemptively when we know that we're running a critical section, like burning a cd, or saving a file to disk.

It would be *really* nice if an app could call a "block" or "unblock" API when the programmer knows in advance where a shutdown should be blocked, releasing the block afterwards.  The "block" API would call `ShutdownBlockReasonCreate(hwnd, L"A critical task is running.");`   The "unblock" API would call `ShutdownBlockReasonDestroy(hwnd);`  The string passed into `ShutdownBlockReasonCreate()` is shown by windows as the reason for the shutdown being held up.

[Application Shutdown Changes in Windows](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/ms700677(v=vs.85))